data/tools/wesnoth/wmltools3.py

```diff
@@ -286,66 +294,126 @@ def isresource(filename):
     return ext and ext[1:] in resource_extensions
 
 def parse_macroref(start, line):
-    def handle_argument():
-        nonlocal opt_arg
-        nonlocal arg
-        nonlocal optional_args
-        nonlocal args
+    def handle_argument(buffer):
+        opt_arg = ""
+
+        arg = "".join(buffer)
+        if arg[0:1].isspace():
+            arg = arg[1:]
 
-        arg = arg.strip()
         # is this an optional argument?
         # argument names are usually made of uppercase letters, numbers and underscores
         # if they're optional, they're followed by an equal sign
         # stop matching on the first one, because the argument value might contain one too
         if re.match(r"^([A-Z0-9_]+?)=", arg):
             opt_arg, arg = arg.split("=", 1)
-        if arg.startswith('"') and arg.endswith('"'):
-            arg = arg[1:-1].strip()
         if opt_arg:
             optional_args[opt_arg] = arg
             opt_arg = ""
         else:
             args.append(arg)
-        arg = ""
+        buffer.clear()
+        return True
 
-    brackdepth = parendepth = 0
-    instring = False
-    optional_args = {}
     args = []
-    opt_arg = ""
-    arg = ""
+    optional_args = {}
+    buffer = []
+
+    depth = {
+        EQUALS: 0,
+        OPEN_BRACE: 0,
+        OPEN_PARENS: 0,
+        QUOTE: 0,
+    }
+    wrapper_stack = []
+    prev = ""
+
+    def close_token(token):
+        while len(wrapper_stack) > 0:
+            last_token = wrapper_stack.pop()
+            depth[last_token] -= 1
+            if last_token == token:
+                break
+        else:
+            return False
+        return True
+
+    def close_if_token(token):
+        if wrapper_stack[-1] == token:
+            wrapper_stack.pop()
+            depth[token] -= 1
+            return True
+        return False
+
+    def open_token(token):
+        wrapper_stack.append(token)
+        depth[token] += 1
+
     for i in range(start, len(line)):
-        if instring:
-            if line[i] == '"':
-                instring = False
-            arg += line[i]
-        elif line[i] == '"':
-            instring = not instring
-            arg += line[i]
-        elif line[i] == "{":
-            if brackdepth > 0:
-                arg += line[i]
-            brackdepth += 1
-        elif line[i] == "}":
-            brackdepth -= 1
-            if brackdepth == 0:
-                if not line[i-1].isspace():
-                    handle_argument()
+        added_arg = False
+        if depth[QUOTE] > 0:
+            if line[i] == QUOTE and line[i+1:i+2] != QUOTE:
```

Comment:
Again slicing: `[i+1:i+2]` is the same as `[i+1]`.

Reply:
Similarly, out of bounds if this is the end of the input.